[DX] Log potential redirect loops caused by forced HTTPS

colinodell opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #27603 
| License       | MIT
| Doc PR        | n/a

If the developer forgets/fails to set "trusted_proxies" properly, forcing the
https channel can cause infinite redirect loops. This change will hopefully
help them identify the problem faster.

See https://github.com/symfony/symfony/issues/27603

colinodell commented:
I have also submitted https://github.com/symfony/symfony-docs/pull/9924 to help point developers in the right direction via documentation.  **That PR does not document this change**, it just provides a different avenue to notify the developer about the potential issue.  Feel free to merge that instead of this (or merge both! :smiley:)

Koc commented on src/Symfony/Component/Security/Http/Firewall/ChannelListener.php:
```diff
@@ -47,6 +47,9 @@ public function handle(GetResponseEvent $event)
             if (null !== $this->logger) {
                 $this->logger->info('Redirecting to HTTPS.');
+                if ('https' === $request->headers->get('X-Forwarded-Proto')) {
+                    $this->logger->debug('Possible redirect loop - did you set "trusted_proxies" correctly?');
```
What about `warn`?

colinodell replied:
This log message could be triggered by anybody who sends HTTP requests to the server with that header:

```bash
curl http://www.example.com -H 'X-Forwarded-Proto: https'
```

If the site is configured to log messages with a minimum severity of `warn` or higher then their logs could potentially be flooded with "false positives".  Using `debug()` would prevent that from happening unless the site is configured to log `debug` messages (which I'd assume somebody might do if they had an unexpected redirect loop and were looking for clues).

That being said, I'm not strongly opposed to using a different log level if others feel that it is more appropriate.

nicolas-grekas replied:
since there is already an info level message, we could merge the new with the existing one:
`$this->logger->info('Redirecting to HTTPS ("X-Forwarded-Proto" is set to "https" - did you set "trusted_proxies" correctly?).');`

colinodell replied:
Good idea!  I've made it so that only one message will ever be logged.

nicolas-grekas commented on src/Symfony/Component/Security/Http/Firewall/ChannelListener.php:
```diff
@@ -47,6 +47,9 @@ public function handle(GetResponseEvent $event)
         if ('https' === $channel && !$request->isSecure()) {
             if (null !== $this->logger) {
                 $this->logger->info('Redirecting to HTTPS.');
+                if ('https' === $request->headers->get('X-Forwarded-Proto')) {
```
this could also handle rfc7239's "proto" parameter

colinodell replied:
I've added an additional check for this parameter.

colinodell committed:
Log potential redirect loops caused by forced HTTPS

If the developer forgets/fails to set "trusted_proxies" properly, forcing the
https channel can cause infinite redirect loops. This change will hopefully
help them identify the problem faster.

See https://github.com/symfony/symfony/issues/27603

fabpot commented:
Thank you @colinodell.
